prow/cmd/deck/job_history.go

```diff
@@ -0,0 +1,291 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/url"
+	"path"
+	"strconv"
+	"strings"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"github.com/sirupsen/logrus"
+	"k8s.io/test-infra/prow/config"
+)
+
+const (
+	resultsPerPage = 20
+	idParam        = "buildId"
+
+	// ** Job history assumes the GCS layout specified here:
+	// https://github.com/kubernetes/test-infra/tree/master/gubernator#gcs-bucket-layout
+	logsPrefix    = "logs"
+	symLinkPrefix = "pr-logs/directory"
+)
+
+type BuildData struct {
+	index        int
+	SpyglassLink string
+	ID           string
+	Started      time.Time
+	Duration     time.Duration
+	Result       string
+}
+
+type JobHistoryTemplate struct {
+	OlderLink  string
+	NewerLink  string
+	LatestLink string
+	Name       string
+	Builds     []BuildData
+}
+
+type started struct {
+	Timestamp int64 `json:"timestamp"`
+}
+
+type finished struct {
+	Timestamp int64  `json:"timestamp"`
+	Result    string `json:"result"`
+}
+
+func readLatestBuild(obj *storage.ObjectHandle) (int, error) {
+	rc, err := obj.NewReader(context.Background())
+	if err != nil {
+		return -1, fmt.Errorf("Failed to get reader for latest build number: %v", err)
+	}
+	data, err := ioutil.ReadAll(rc)
+	if err != nil {
+		return -1, fmt.Errorf("Failed to read latest build number: %v", err)
+	}
+	n64, err := strconv.Atoi(string(data))
+	if err != nil {
+		return -1, fmt.Errorf("Failed to parse latest build number: %v", err)
+	}
+	return int(n64), nil
+}
+
+// resolve sym links into the actual log directory for a particular test run
+func resolveSymLink(bkt *storage.BucketHandle, symLink string) (string, error) {
+	linkObj := bkt.Object(symLink)
+	rc, err := linkObj.NewReader(context.Background())
+	if err != nil {
+		return "", err
+	}
+	data, err := ioutil.ReadAll(rc)
+	if err != nil {
+		return "", err
+	}
+	bAttrs, err := bkt.Attrs(context.Background())
+	if err != nil {
+		return "", err
+	}
+	// strip gs://<bucket-name> from global address `u`
+	u := string(data)
+	i := strings.Index(u, bAttrs.Name)
+	dir := u[i+len(bAttrs.Name)+1:] // +1 for leading slash
+	return dir, nil
+}
+
+func getDir(bkt *storage.BucketHandle, root, id string) string {
+	bAttrs, err := bkt.Attrs(context.Background())
+	if err != nil {
+		return ""
```

Comment:
This function should return `(string, error)` so that we don't throw out the error messages (the caller should handle and log them).